crates/engine_bibtex/src/c_api/char_info.rs

```diff
@@ -0,0 +1,217 @@
+    };
+    ($start:literal..=$end:literal => $arr:ident[..] = $expr:expr) => {
+        let mut i = $start;
+        while i <= $end {
+            $arr[i] = $expr;
+            i += 1;
+        }
+    };
+}
+
+/// The lexer class of a character - this represents whether the parser considers it to be alphabetic,
+/// numeric, etc. Illegal represents tokens that shouldn't show up at all, such as ASCII backspace.
+///
+/// cbindgen:rename-all=ScreamingSnakeCase
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[repr(C)]
+pub enum LexClass {
+    Illegal = 0,
+    Whitespace = 1,
+    Alpha = 2,
+    Numeric = 3,
+    Sep = 4,
+    Other = 5,
+}
+
+impl LexClass {
+    /// Get the `LexClass` of a character
+    pub fn of(char: ASCIICode) -> LexClass {
+        LEX_CLASS[char as usize]
+    }
+}
+
+/// Whether a token counts as valid in an identifier
+///
+/// cbindgen:rename-all=ScreamingSnakeCase
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[repr(C)]
+pub enum IdClass {
+    IllegalIdChar = 0,
+    LegalIdChar = 1,
+}
+
+impl IdClass {
+    /// Get the `IdClass` of a character
+    pub const fn of(char: ASCIICode) -> IdClass {
+        ID_CLASS[char as usize]
+    }
+}
+
+#[no_mangle]
+pub static LEX_CLASS: [LexClass; 256] = {
+    let mut lex_class = [LexClass::Other; 256];
+
+    const_for!(128..=255 => lex_class[..] = LexClass::Alpha);
+    const_for!(0..32 => lex_class[..] = LexClass::Illegal);
+
+    // invalid_code
+    lex_class[127] = LexClass::Illegal;
+    // tab
+    lex_class[9] = LexClass::Whitespace;
+    lex_class[13] = LexClass::Whitespace;
+    // space
+    lex_class[32] = LexClass::Whitespace;
+    // tie
+    lex_class[126] = LexClass::Sep;
+    // hyphen
+    lex_class[45] = LexClass::Sep;
+
+    const_for!(48..58 => lex_class[..] = LexClass::Numeric);
+    const_for!(65..91 => lex_class[..] = LexClass::Alpha);
+    const_for!(97..123 => lex_class[..] = LexClass::Alpha);
+
+    lex_class
+};
+
+pub const ID_CLASS: [IdClass; 256] = {
+    let mut id_class = [IdClass::LegalIdChar; 256];
+
+    const_for!(0..32 => id_class[..] = IdClass::IllegalIdChar);
+    // tab
+    id_class[9] = IdClass::IllegalIdChar;
+    // space
+    id_class[32] = IdClass::IllegalIdChar;
+    // double quote
+    id_class[34] = IdClass::IllegalIdChar;
+    // number sign
+    id_class[35] = IdClass::IllegalIdChar;
+    // comment
+    id_class[37] = IdClass::IllegalIdChar;
+    // single quote
+    id_class[39] = IdClass::IllegalIdChar;
+    // left paren
+    id_class[40] = IdClass::IllegalIdChar;
+    // right paren
+    id_class[41] = IdClass::IllegalIdChar;
+    // comma
+    id_class[44] = IdClass::IllegalIdChar;
+    // equals sign
+    id_class[61] = IdClass::IllegalIdChar;
+    // left brace
+    id_class[123] = IdClass::IllegalIdChar;
+    // right brace
+    id_class[125] = IdClass::IllegalIdChar;
+
+    id_class
+};
+
+#[no_mangle]
+pub static CHAR_WIDTH: [i32; 256] = {
+    let mut char_width = [0; 256];
+
```

Comment:
It might be nice to have a comment here explaining where these magical values come from (even if it's just "these magical values come from the bibtex web2c implementation").

Reply:
Adding a comment for this, and also fixing it up to use the `b''` character syntax in most places - lets me remove some comments, and generally makes the intent a lot more obvious.